snapcraft/internal/meta/command.py

```diff
@@ -43,62 +37,174 @@
 )
 
 
-def _get_shebang_from_file(file_path: str) -> List[str]:
+def _get_shebang_from_file(file_path: str) -> Optional[str]:
     """Returns the shebang from file_path."""
-    if not os.path.exists(file_path):
-        raise errors.ShebangNotFoundError()
-
     with open(file_path, "rb") as exefile:
         if exefile.read(2) != b"#!":
-            raise errors.ShebangNotFoundError()
+            return None
         shebang_line = exefile.readline().strip().decode("utf-8")
 
-    # posix is set to False to respect the quoting of variables.
-    shebang_parts = shlex.split(shebang_line, posix=False)
-    # remove the leading /usr/bin/env.
-    if shebang_parts[0] == "/usr/bin/env":
-        shebang_parts = shebang_parts[1:]
-    # or if the shebang startswith /.
-    elif shebang_parts[0].startswith("/"):
-        raise errors.ShebangInRoot()
+    return shebang_line
 
-    return shebang_parts
 
+def _find_executable(*, command: str, prime_dir: str) -> Optional[str]:
+    """Find executable in common binary path locations.
 
-def _find_executable(*, binary: str, prime_dir: str) -> str:
-    found_path: Optional[str] = None
+    :returns: path to executable found, whether it is in searched
+              prime directory paths, or on host using shutil.which().
+    """
     binary_paths = (
-        os.path.join(p, binary) for p in common.get_bin_paths(root=prime_dir)
+        os.path.join(p, command) for p in common.get_bin_paths(root=prime_dir)
     )
     for binary_path in binary_paths:
         if _executable_is_valid(binary_path):
-            found_path = binary_path
-            break
-    else:
-        # Last chance to find in the prime_dir, mostly for backwards compatibility,
-        # to find the executable, historical snaps like those built with the catkin
-        # plugin will have roslaunch in a path like /opt/ros/bin/roslaunch.
-        for root, _, files in os.walk(prime_dir):
-            if _executable_is_valid(os.path.join(root, binary)):
-                found_path = os.path.join(root, binary)
-                break
+            return binary_path
+
+    # Last chance to find in the prime_dir, mostly for backwards compatibility,
+    # to find the executable, historical snaps like those built with the catkin
+    # plugin will have roslaunch in a path like /opt/ros/bin/roslaunch.
+    for root, _, files in os.walk(prime_dir):
+        if _executable_is_valid(os.path.join(root, command)):
+            return os.path.join(root, command)
+
+    # Finally, check if it is part of the system.
+    return shutil.which(command)
+
+
+def _resolve_snap_command_path(
+    *, command: str, prime_dir: str
```

Comment:
The reason I originally went with binary instead of command was to easily untangle the entire command entry and the actual command.

Can you better explain what is done here? Is this supposed to work with $SNAP/foo or is that to be resolved in a different step? What is a typical binary path?

Reply:
Let me know if the refactor clears this up.